mailpoet/lib/Subscribers/ImportExport/Import/Import.php

```diff
@@ -619,4 +634,42 @@ public function addSubscribersToSegments(array $subscribersIds, array $segmentsI
       );
     }
   }
+
+  /**
+   * @param int[] $subscribersIds
+   * @param string[] $tagNames
+   */
+  public function addTagsToSubscribers(array $subscribersIds, array $tagNames): void {
+    $tagIds = [];
+    foreach ($tagNames as $tagName) {
+      $tag = $this->tagRepository->findOneBy(['name' => $tagName]);
+      if (!$tag) {
+        $tag = $this->tagRepository->createOrUpdate(['name' => $tagName]);
+      }
+      $tagIds[] = $tag->getId();
+    }
+
+    $columns = [
+      'subscriber_id',
+      'tag_id',
+      'created_at',
+    ];
+    foreach (array_chunk($subscribersIds, self::DB_QUERY_CHUNK_SIZE) as $subscriberIdsChunk) {
+      $data = [];
+      foreach ($tagIds as $tagId) {
+        $data = array_merge($data, array_map(function ($subscriberId) use ($tagId): array {
+          return [
+            $subscriberId,
+            $tagId,
+            $this->createdAt,
+          ];
+        }, $subscriberIdsChunk));
+      }
+      $this->importExportRepository->insertMultiple(
+        SubscriberTagEntity::class,
+        $columns,
+        $data
+      );
```

Comment:
I see that we use multiple inserts in batches and we split them into chunks by splitting subscribers. I just want to point out that because of the nested for each we may actually generate multiple insert with more than self::DB_QUERY_CHUNK_SIZE items.
The chunk size is 100. 100 subscribers each with 100 tags will generate multiple inserts with 10000 rows.
This is probably not a big deal since there most probably won't be that many tags. If we want to be strict we modify the script so that it counts the number of generated rows and flushes them to DB each time they reach 100 and then last time at the end.

Reply:
@costasovo I think you overlooked the custom solution for [inserting multiple rows in the importExportRepository](https://github.com/mailpoet/mailpoet/blob/trunk/mailpoet/lib/Subscribers/ImportExport/ImportExportRepository.php#L74-L102). This custom solution inserts all tags for the chunk size with one query. That means that 100 subscribers with 100 tags are inserted by one multi-insert.
I tried to test this solution on my local, and 1000 subscribers with 5 tags were stored by 10 queries.

Reply:
> I tried to test this solution on my local, and 1000 subscribers with 5 tags were stored by 10 queries.

@lysyjan I think that exactly this might potentially be the issue. Our constant self::DB_QUERY_CHUNK_SIZE is set to 100. My interpretation of the constant (please correct me if I'm wrong) is that we want to limit the number of rows in one multi-insert to 100.  Importing 1000 subscribers with 5 tags each in 10 queries means we had 500 rows in a single multi-insert query which is five times more than the chunk size defined by the constant. So the current implementation allows bigger chunks than defined in the constant and theoretically can lead to huge multi-insert queries.
I think that the expected behavior should be 1000 subscribers with 5 tags = 5000 rows -> so we need 50 queries (100 rows per query).

This may cause an issue only in some extreme cases like subscribers with hundreds of tags. The potential issue with such huge inserts could be that if SQL query becomes bigger than `max_allowed_packet`, MySQL will throw an error. https://www.mysqltutorial.org/mysql-insert-multiple-rows/

Reply:
@costasovo Thank you for your explanation. It makes sense to me now. We have the same behavior for the segments. So, I would suggest changing this behavior for both cases. I could change the `foreach` order so that each query inserts only the exact count of records. Do you agree with my proposal?

Reply:
I haven't realized that it is doable by changing the order of the loops. If so, yes I agree.
An alternative solution could be storing rows in an array and each time it has 100 items we do the query + one query at the end.

Reply:
@costasovo I changed it and added a commit with another fix.